embedded/xsens_imu/src/imu.c

```diff
@@ -0,0 +1,216 @@
+/*
+* imu.c
+* Author: Rohan Daruwala
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include <pthread.h>
+#include <semaphore.h> 
+#include <unistd.h>
+#include "imu.h"
+#include "i2c.h"
+
+//Global Variables
+IMU_data * data;
+i2c_settings * i2c;
+pthread_t IMUThread;
+sem_t vMutex, oMutex;
+
+void SetupIMU(){
+	sem_init(&vMutex, 0, 1);
+	sem_init(&oMutex, 0, 1);
+	i2c = malloc(sizeof(i2c_settings));
+	i2c->bus = 2;
+	i2c->deviceAddress = I2C_ADDRESS;
+	i2c->openMode = O_RDWR;
+
+	data = malloc(sizeof(IMU_data));
+
+	if (i2c_begin(i2c) == -1){
+		fprintf(stderr, "Could not open i2c bus.\n");
+		return;
+	}
+
+	if (pthread_create(&IMUThread, NULL, IMULoop, NULL)){
+		fprintf(stderr, "Error creating IMU thread\n");
+	}
+
+	//Set all IMU structure values to zero
+	data->dVx = 0;
+	data->dVy = 0;
+	data->dVz = 0;
+
+	data->accelX = 0;
+	data->accelY = 0;
+	data->accelZ = 0;
+
+	data->q0 = 0;
+	data->q1 = 0;
+	data->q2 = 0;
+	data->q3 = 0;
+}
+
+void *IMULoop(){
+
+	while (1){
+		
+		//Get message length size
+		write_byte_i2c(i2c, STATUS_REG);
+		unsigned char res1[4];
+		read_i2c(i2c, res1, 4);
+
+		uint16_t mSize = res1[2] | res1[3] << 8;
+
+
+		// Get Data
+		write_byte_i2c(i2c, DATA_REG);
+		unsigned char dataBuffer[mSize];
+		read_i2c(i2c, dataBuffer, mSize);
+
+		for (int i = 0; i < mSize; i++){
```

Comment:
You may want to `sem_wait` at the beginning of this loop and `sem_post` at the end instead of specifically for each burst of writes.

This loop is also confusing, `for (int i = 0; i < mSize; i++)` expresses that we are doing some kind of work on the message for each byte, but really we're looking forward into the packet for some kind of header/fingerprint that tells us what the data is and then processing it in place.

Would it be possible to set up the loop so its more clear the data is being processed in chunks?